Update readme and changes log for v1.0.2 (#17)
SiliconLabsSoftware/z-wave-nvm-migration-tool: README.md

Reviewer: Please use a formal and impersonal writing style to avoid 'you/we/etc.'

# Z-WAVE NVM TOOL 
The Z-Wave NVM tool simplifies the process of reading an NVM3 image and converting it into a human-readable JSON format. It also provides functionality to upgrade NVM3 files to newer versions, facilitating controller migration while preserving Z-Wave network information.
****

## Build Instructions
### Testing Environment: 
- Linux system: (Debian-12, Ubuntu-22.04+ to 24.04+) on amd64 , armhf architecture
- Hardware: ZGM230S module (BRD4207A) + WSTK (BRD4002A)

> **_Note:_** The following commands are executed on the Ubuntu terminal.

> ### Installation

Reviewer 2: May you add a chapter about download released deb files for various platform

````
sudo apt-get update
sudo dpkg -i ./*.deb \
  || sudo apt install -f -y --no-install-recommends 
````

See https://github.com/SiliconLabsSoftware/z-wave-protocol-controller/blob/main/Dockerfile#L45

Author: Will the release packages for various platforms be delivered when we release a tag on: https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/tags, right?

Reviewer 2: yes

but you can link to

https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/releases

there we will display release notes

Author: Yes, I've edited that to point to [release page](https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/releases)

# Z-WAVE NVM TOOL 
The Z-Wave NVM tool simplifies the process of reading an NVM3 image and converting it into a human-readable JSON format. It also provides functionality to upgrade NVM3 files to newer versions, facilitating controller migration while preserving Z-Wave network information.
****

## Build Instructions
### Testing Environment: 
- Linux system: (Debian-12, Ubuntu-22.04+ to 24.04+) on amd64 , armhf architecture
- Hardware: ZGM230S module (BRD4207A) + WSTK (BRD4002A)

> **_Note:_** The following commands are executed on the Ubuntu terminal.
### Install from release packages
Download the release packages for various platforms from the [Release page](https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/releases).

> Install the `.deb` package on your system using the following commands:

Reviewer 2: ```suggestion
Install the `.deb` package (matching your system architecture) using the following commands:
```

```sh
sudo apt-get update
sudo dpkg -i ./*.deb \
  || sudo apt install -f -y --no-install-recommends 

> ```

Reviewer 2: The application can be run from path and it can uses the shema file from /usr/share/doc/z-wave-nvm-migration-tool/...

check z-wave-nvm-migration-tool usage

cc: @vosavost

### Build from source
The project is CMake based, to prepare the environment, have a look at [./helper.mk](./helper.mk)'s details for needed steps to setup developer system before using CMake normally.

At the moment stable version of Debian (12) is supported, so it should work also in relatives projects (Ubuntu, RaspiOS, WSL2 etc) and should be easy to adapt to other distributions.

**Requirements**: 
Install dependencies 
```sh
$ sudo apt install make git sudo
$ ./helper.mk setup 
```
> Note: json-c requires cmake from v3.9. To build with "Debug mode", please edit 
```option(ENABLE_DEBUG "Debug Mode is ON/OFF" ON)```

**Download and build process**
- Download the .zip file 
- Unzip the file
- Build via steps below
```sh
$ ./helper.mk VERBOSE=1 # Default build tasks verbosely (depends on setup)
```
An executable will be created in build folder. 
```
z-wave-nvm-migration-tool/
├── CHANGES.md
├── CMakeLists.txt
├── Dockerfile
├── LICENSE
├── README.md
├── build
├── cmake
├── extra
├── helper.mk
├── sonar-project.properties
├── tests
├── zw_nvm_converter
├── zwave_data_description_scheme.json
├── zwave_nvm_tool.c
└── zwave_nvm_tool_version.h
```

> > **_Note:_** The pre-built binary is placed in pre_built folder.

Reviewer 2: ```suggestion
> **_Note:_** The built binary is placed in build folder (if not installed to system, path need to be adjusted in following commands)
```

> **_Note:_** `zwave_data_description_scheme.json` is the schema file containing description of all properties of Z-Wave objects that exist in a NVM3 file and is used for the migration process.

## Usage
```sh

> Z-Wave NVM3 Migration Tool Version 1.0.2

Reviewer 2: ```suggestion
$ z-wave-nvm-migration-tool
Z-Wave NVM3 Migration Tool Version
```

It will be future ready

Usage: ./z-wave-nvm-migration-tool [options] <arguments>
Options:
    -e  <nvm_file_input>
        Convert: READ an NVM3 file and output data in JSON format.

>         Example: ./z-wave-nvm-migration-tool -e input.bin

Reviewer 2: ```suggestion
        Example: z-wave-nvm-migration-tool -e input.bin
```

    -i <json_file_input> <part_number>
        Convert: Parse data from a JSON file (aligned with schema) and generate an NVM3 file for the specified hardware.
        Supported parts: EFR32XG13, EFR32XG14, EFR32XG23, EFR32XG28

>         Example: ./z-wave-nvm-migration-tool -i input.json EFR32XG23

Reviewer 2: ```suggestion
        Example: z-wave-nvm-migration-tool -i input.json EFR32XG23
```

Reviewer 2: Where is input.json coming from ?

    -u <json_file_input> <target_protocol_version> <schema_file>
        Upgrade: Upgrade a JSON data file to a newer version using a schema.
        If target_protocol_version == current_protocol_version: populate missing keys with default values.

>         Example: ./z-wave-nvm-migration-tool -u input.json 7.19.0 zwave_data_description_scheme.json

Reviewer 2: ```suggestion
        Example: z-wave-nvm-migration-tool -u input.json 7.19.0 zwave_data_description_scheme.json
```

```

### Use case 2: Adding missing data for a version 7.21.0 for 700 Series (XG13/XG14) 
NVM3 images generated by 700 Series controllers with firmware version 7.21.0 will lack the application data object. For instance, consider the file `nvm_7_21_0.bin`.

This file can be updated either by manual modification or by utilizing the `-u` option to add missing objects according to the schema.

#### 1. Manually (to inspect the data in the NVM3 file): 

**Step 1**:	Read data from `nvm_7_21_0.bin` and save to json file

```sh
$ ./z-wave-nvm-migration-tool -e nvm_7_21_0.bin -o nvm_7_21_0.json
```

**Step 2**: Populate missing keys using the `-u` option:

```sh
$ ./z-wave-nvm-migration-tool -u nvm_7_21_0.json 7.21.0 zwave_data_description_scheme.json -o nvm_7_21_0_adding_missing_keys.json
```

**Step 3**: Using the newly upgraded JSON data file (7.21.0) to generate a NVM3 file
```sh
$ ./z-wave-nvm-migration-tool -i nvm_7_21_0_adding_missing_keys.json EFR32XG23 -o nvm_7_21_0_adding_missing_keys.bin 
```
- NVM3 data file has been generated for `EFR32XG23` 

#### 2. Using option `-m`: 
```sh

> $ ./z-wave-nvm-migration-tool -m nvm_7_21_0.bin 7.21.0 EFR32XG13 zwave_data_description_scheme.json -o nvm_7_21_0_adding_missing_keys.bin 

Reviewer: In case of *.deb usage, I have no idea where to find JSON-scheme file.
And it would be nice to have it by default

Reviewer 2: then it should be packaged , i can do that

btw  please @DucAnh-silabs  you can remove ./ prefix if it is installed it will be resolved from $PATH

# Z-WAVE NVM TOOL 
The Z-Wave NVM tool simplifies the process of reading an NVM3 image and converting it into a human-readable JSON format. It also provides functionality to upgrade NVM3 files to newer versions, facilitating controller migration while preserving Z-Wave network information.
****

## Build Instructions
### Testing Environment: 

> - PC: Ubuntu 18.04.5 LTS (Bionic Beaver), 22.04.1 LTS (Jammy Jellyfish) and 24.04.1 LTS (Noble Numbat)

Reviewer 2: ```suggestion
- Linux system: (Debian-12, Ubuntu-22.04+ to 24.04+) on amd64 , armhf architecture
```

> - Hardware: ZGM230S module (BRD4207A) + WSTK (BRD4002A)

Reviewer 2: May we add firmware filenames and link to:

https://github.com/SiliconLabs/simplicity_sdk/releases/download/v2025.6.1/demo-applications.zip

Reviewer 2: can be done in later version

> **_Note:_** The following commands are executed on the Ubuntu terminal.
### Installation
**Requirements**: 
Install dependencies 
```sh

> $ ./helper.mk setup 

Reviewer 2: ```suggestion
$ sudo apt install make git sudo
$ ./helper.mk setup 
```

        Supported parts: EFR32XG13, EFR32XG14, EFR32XG23, EFR32XG28
        Example: ./z-wave-nvm-migration-tool -i input.json EFR32XG23

    -u <json_file_input> <target_protocol_version> <schema_file>
        Upgrade: Upgrade a JSON data file to a newer version using a schema.
        If target_protocol_version == current_protocol_version: populate missing keys with default values.
        Example: ./z-wave-nvm-migration-tool -u input.json 7.19.0 zwave_data_description_scheme.json

    -m <nvm_file_input> <target_protocol_version> <part_number> <schema_file>
        Migration: Update an NVM3 file to a new version using a schema.
        If target_protocol_version == current_protocol_version: populate missing keys with default values.
        Example: ./z-wave-nvm-migration-tool -m input.bin 7.19.0 EFR32XG23 schema.json

    -o <output_file>
        Specify an output file; otherwise, the output file name will be generated automatically.
        This option can be used along with other options.
        Example: ./z-wave-nvm-migration-tool -e input.bin -o output.json

    -h
        Display this help message
```

## Some Specific Use Cases 

> ### Use case #1: Migrating NVM3 for restoring on a different controller

Reviewer 2: ```suggestion
### Use case 1: Migrating NVM3 for restoring on a different controller
```

#### 1. Manually (if you want to inspect the data in the NVM3 file):

> - Assume that we have the NVM3 image `nvm_7_18.bin` that is read from one controller version 7.18.0 using either `commander` or serialAPI using `zw_programmer`

Reviewer 2: Maybe we can explain how to dump file from flashed firmware (can automate that in helper.mk too)

Author: I'll add example about `commander` and `zw_programmer` usages

Author: Oh wait, I've already added them, it's down below

Author: But I'll link to it

Reviewer 2: i have code to borrow about dl tools at:
 https://github.com/Z-Wave-Alliance/z-wave-stack/blob/236d20c0d36a261608b135b6d448f71389ee04b2/platform/SiliconLabs/helper.mk

I can add that for later version

- Unzip the file
- Build via steps below
```sh

> $ ./helper.mk VERBOSE=1 # Default build tasks verbosely (depends on setup)

Reviewer: Add helper description, please.

Reviewer 2: you can duplicate to what i have said in :

https://github.com/SiliconLabsSoftware/z-wave-protocol-controller/blob/main/README.md

The project is CMake based, to prepare the environment, have a look at [./helper.mk](https://github.com/SiliconLabsSoftware/z-wave-protocol-controller/blob/main/helper.mk)'s details for needed steps to setup developer system before using CMake normally.

At the moment stable version of Debian (12) is supported, so it should work also in relatives projects (Ubuntu, RaspiOS, WSL2 etc) and should be easy to adapt to other distributions.

```
An executable will be created in build folder. 
```
z-wave-nvm-migration-tool/
├── CHANGES.md
├── CMakeLists.txt
├── Dockerfile
├── LICENSE
├── README.md
├── build
├── cmake
├── extra
├── helper.mk
├── sonar-project.properties
├── tests
├── zw_nvm_converter
├── zwave_data_description_scheme.json
├── zwave_nvm_tool.c
└── zwave_nvm_tool_version.h
```
> **_Note:_** The pre-built binary is placed in pre_built folder.

> **_Note:_** `zwave_data_description_scheme.json` is the schema file containing description of all properties of Z-Wave objects that exist in a NVM3 file and is used for the migration process.

## Usage
```sh
Z-Wave NVM3 Migration Tool Version 1.0.2

> Usage: ./z-wave-nvm-migration-tool [options] <arguments>

Reviewer: Please highlight this change - it is not obvious

Reviewer 2: version can be removed for future
